pkg/distributor/distributor.go

```diff
@@ -744,7 +744,27 @@ func (d *Distributor) Push(ctx context.Context, req *cortexpb.WriteRequest) (*co
 		// Return a 429 here to tell the client it is going too fast.
 		// Client may discard the data or slow down and re-send.
 		// Prometheus v2.26 added a remote-write option 'retry_on_http_429'.
-		return nil, httpgrpc.Errorf(http.StatusTooManyRequests, "ingestion rate limit (%v) exceeded while adding %d samples and %d metadata", d.ingestionRateLimiter.Limit(now, userID), validatedSamples, len(validatedMetadata))
+		var rateLimitErr limiter.RateLimitError
+		if errors.As(err, &rateLimitErr) {
+			if rateLimitErr.Retryable {
+				return nil, httpgrpc.ErrorFromHTTPResponse(&httpgrpc.HTTPResponse{
+					Code: http.StatusTooManyRequests,
+					Headers: []*httpgrpc.Header{{
+						Key:    "Retry-After",
+						Values: []string{rateLimitErr.RetryAfter},
+					}},
+					Body: []byte(fmt.Sprintf("ingestion rate limit (%v) exceeded while adding %d samples and %d metadata, your may retry this request later with Retry-After(%s)",
+						rateLimitErr.Limit, validatedSamples, len(validatedMetadata), rateLimitErr.RetryAfter)),
+				})
+			}
+			return nil, httpgrpc.ErrorFromHTTPResponse(&httpgrpc.HTTPResponse{
+				Code: http.StatusRequestEntityTooLarge,
+				Body: []byte(fmt.Sprintf("ingestion rate limit burst (%v) exceeded while adding %d samples and %d metadata, please break this request into multiple small requests",
+					rateLimitErr.Burst, validatedSamples, len(validatedMetadata))),
+			})
+		}
+		return nil, httpgrpc.Errorf(http.StatusTooManyRequests, "ingestion rate limit (%v) exceeded while adding %d samples and %d metadata",
```

Comment:
Is there anyway we can avoid having the need of this line of code? Would be nice if we only have to deal with `limiter.RateLimitError` type.